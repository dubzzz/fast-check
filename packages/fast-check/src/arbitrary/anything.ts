import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { anyArbitraryBuilder } from './_internals/builders/AnyArbitraryBuilder';
import type { ObjectConstraints } from './_internals/helpers/QualifiedObjectConstraints';
import { toQualifiedObjectConstraints } from './_internals/helpers/QualifiedObjectConstraints';

export type { ObjectConstraints };

/**
 * For any type of values
 *
 * You may use {@link sample} to preview the values that will be generated
 *
 * @example
 * ```javascript
 * null, undefined, 42, 6.5, 'Hello', {}, {k: [{}, 1, 2]}
 * ```
 *
 * @remarks Since 0.0.7
 * @public
 */
function anything(): Arbitrary<unknown>;
/**
 * For any type of values following the constraints defined by `settings`
 *
 * You may use {@link sample} to preview the values that will be generated
 *
 * @example
 * ```javascript
 * null, undefined, 42, 6.5, 'Hello', {}, {k: [{}, 1, 2]}
 * ```
 *
 * @example
 * ```typescript
 * // Using custom settings
 * fc.anything({
 *     key: fc.char(),
 *     values: [fc.integer(10,20), fc.constant(42)],
 *     maxDepth: 2
 * });
 * // Can build entries such as:
 * // - 19
 * // - [{"2":12,"k":15,"A":42}]
 * // - {"4":[19,13,14,14,42,11,20,11],"6":42,"7":16,"L":10,"'":[20,11],"e":[42,20,42,14,13,17]}
 * // - [42,42,42]...
 * ```
 *
 * @param constraints - Constraints to apply when building instances
 *
 * @remarks Since 0.0.7
 * @public
 */
function anything(constraints: ObjectConstraints): Arbitrary<unknown>;
function anything(constraints?: ObjectConstraints): Arbitrary<unknown> {
  return anyArbitraryBuilder(toQualifiedObjectConstraints(constraints));
}
export { anything };
