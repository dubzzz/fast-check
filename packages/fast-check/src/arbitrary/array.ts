import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { ArrayArbitrary } from './_internals/ArrayArbitrary';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import {
  MaxLengthUpperBound,
  maxGeneratedLengthFromSizeForArbitrary,
} from './_internals/helpers/MaxLengthFromMinLength';
import type { DepthIdentifier } from './_internals/helpers/DepthContext';

/**
 * Constraints to be applied on {@link array}
 * @remarks Since 2.4.0
 * @public
 */
export interface ArrayConstraints {
  /**
   * Lower bound of the generated array size
   * @defaultValue 0
   * @remarks Since 2.4.0
   */
  minLength?: number;
  /**
   * Upper bound of the generated array size
   * @defaultValue 0x7fffffff â€” _defaulting seen as "max non specified" when `defaultSizeToMaxWhenMaxSpecified=true`_
   * @remarks Since 2.4.0
   */
  maxLength?: number;
  /**
   * Define how large the generated values should be (at max)
   *
   * When used in conjonction with `maxLength`, `size` will be used to define
   * the upper bound of the generated array size while `maxLength` will be used
   * to define and document the general maximal length allowed for this case.
   *
   * @remarks Since 2.22.0
   */
  size?: SizeForArbitrary;
  /**
   * When receiving a depth identifier, the arbitrary will impact the depth
   * attached to it to avoid going too deep if it already generated lots of items.
   *
   * In other words, if the number of generated values within the collection is large
   * then the generated items will tend to be less deep to avoid creating structures a lot
   * larger than expected.
   *
   * For the moment, the depth is not taken into account to compute the number of items to
   * define for a precise generate call of the array. Just applied onto eligible items.
   *
   * @remarks Since 2.25.0
   */
  depthIdentifier?: DepthIdentifier | string;
}

/**
 * Extra but internal constraints that can be passed to array.
 * This extra set is made of constraints mostly targets experimental and internal features not yet mature to be exposed.
 * @internal
 */
export interface ArrayConstraintsInternal<T> extends ArrayConstraints {
  /**
   * Extra user-definable and hardcoded values.
   * Each entry in the array could be used to build the final generated value outputed by the arbitrary of array on generate.
   * Each entry must have at least one element of type T into it.
   * Each T must be a value acceptable for the arbitrary passed to the array.
   */
  experimentalCustomSlices?: T[][];
}

/**
 * For arrays of values coming from `arb`
 *
 * @param arb - Arbitrary used to generate the values inside the array
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
function array<T>(arb: Arbitrary<T>, constraints: ArrayConstraints = {}): Arbitrary<T[]> {
  const size = constraints.size;
  const minLength = constraints.minLength || 0;
  const maxLengthOrUnset = constraints.maxLength;
  const depthIdentifier = constraints.depthIdentifier;
  const maxLength = maxLengthOrUnset !== undefined ? maxLengthOrUnset : MaxLengthUpperBound;
  const specifiedMaxLength = maxLengthOrUnset !== undefined;
  const maxGeneratedLength = maxGeneratedLengthFromSizeForArbitrary(size, minLength, maxLength, specifiedMaxLength);
  const customSlices = (constraints as ArrayConstraintsInternal<T>).experimentalCustomSlices || [];
  return new ArrayArbitrary<T>(arb, minLength, maxGeneratedLength, maxLength, depthIdentifier, undefined, customSlices);
}
export { array };
