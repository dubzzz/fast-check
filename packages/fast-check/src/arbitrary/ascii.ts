import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildCharacterArbitrary } from './_internals/builders/CharacterArbitraryBuilder';
import { indexToPrintableIndexMapper, indexToPrintableIndexUnmapper } from './_internals/mappers/IndexToPrintableIndex';

/**
 * For single ascii characters - char code between 0x00 (included) and 0x7f (included)
 * @remarks Since 0.0.1
 * @public
 */
export function ascii(): Arbitrary<string> {
  return buildCharacterArbitrary(0x00, 0x7f, indexToPrintableIndexMapper, indexToPrintableIndexUnmapper);
}
