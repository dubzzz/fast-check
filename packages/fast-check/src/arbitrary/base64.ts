import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildCharacterArbitrary } from './_internals/builders/CharacterArbitraryBuilder';

/** @internal */
function base64Mapper(v: number) {
  if (v < 26) return v + 65; // A-Z
  if (v < 52) return v + 97 - 26; // a-z
  if (v < 62) return v + 48 - 52; // 0-9
  return v === 62 ? 43 : 47; // +/
}

/** @internal */
function base64Unmapper(v: number) {
  if (v >= 65 && v <= 90) return v - 65; // A-Z
  if (v >= 97 && v <= 122) return v - 97 + 26; // a-z
  if (v >= 48 && v <= 57) return v - 48 + 52; // 0-9
  return v === 43 ? 62 : v === 47 ? 63 : -1; // +/
}

/**
 * For single base64 characters - A-Z, a-z, 0-9, + or /
 * @remarks Since 0.0.1
 * @public
 */
export function base64(): Arbitrary<string> {
  return buildCharacterArbitrary(0, 63, base64Mapper, base64Unmapper);
}
