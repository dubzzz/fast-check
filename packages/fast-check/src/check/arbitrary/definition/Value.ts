import type { WithCloneMethod } from '../../symbols';
import { cloneMethod, hasCloneMethod } from '../../symbols';

const safeObjectDefineProperty = Object.defineProperty;

/**
 * A `Value<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 3.0.0 (previously called `NextValue` in 2.15.0)
 * @public
 */
export class Value<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = customGetValue !== undefined || hasCloneMethod(value_);
    this.readOnce = false;
    if (this.hasToBeCloned) {
      safeObjectDefineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
    } else {
      this.value = value_;
    }
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return (this.value_ as unknown as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
