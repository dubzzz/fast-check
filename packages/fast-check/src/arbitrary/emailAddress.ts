import { array } from './array';
import { buildLowerAlphaNumericArbitrary } from './_internals/builders/CharacterRangeArbitraryBuilder';
import { domain } from './domain';
import { stringOf } from './stringOf';
import { tuple } from './tuple';
import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import type { AdapterOutput } from './_internals/AdapterArbitrary';
import { adapter } from './_internals/AdapterArbitrary';
import { safeJoin, safeSlice, safeSplit } from '../utils/globals';

/** @internal */
function dotAdapter(a: string[]): AdapterOutput<string[]> {
  // According to RFC 2821:
  //    The maximum total length of a user name or other local-part is 64 characters.
  let currentLength = a[0].length; // always at least one element
  for (let index = 1; index !== a.length; ++index) {
    currentLength += 1 + a[index].length;
    if (currentLength > 64) {
      return { adapted: true, value: safeSlice(a, 0, index) };
    }
  }
  return { adapted: false, value: a };
}
/** @internal */
function dotMapper(a: string[]): string {
  return safeJoin(a, '.');
}
/** @internal */
function dotUnmapper(value: unknown): string[] {
  if (typeof value !== 'string') {
    throw new Error('Unsupported');
  }
  return safeSplit(value, '.');
}
/** @internal */
function atMapper(data: [string, string]): string {
  return `${data[0]}@${data[1]}`;
}
/** @internal */
function atUnmapper(value: unknown): [string, string] {
  if (typeof value !== 'string') {
    throw new Error('Unsupported');
  }
  return safeSplit(value, '@', 2) as [string, string];
}

/**
 * Constraints to be applied on {@link emailAddress}
 * @remarks Since 2.22.0
 * @public
 */
export interface EmailAddressConstraints {
  /**
   * Define how large the generated values should be (at max)
   * @remarks Since 2.22.0
   */
  size?: Exclude<SizeForArbitrary, 'max'>;
}

/**
 * For email address
 *
 * According to {@link https://www.ietf.org/rfc/rfc2821.txt | RFC 2821},
 * {@link https://www.ietf.org/rfc/rfc3696.txt | RFC 3696} and
 * {@link https://www.ietf.org/rfc/rfc5322.txt | RFC 5322}
 *
 * @param constraints - Constraints to apply when building instances (since 2.22.0)
 *
 * @remarks Since 1.14.0
 * @public
 */
export function emailAddress(constraints: EmailAddressConstraints = {}): Arbitrary<string> {
  const others = ['!', '#', '$', '%', '&', "'", '*', '+', '-', '/', '=', '?', '^', '_', '`', '{', '|', '}', '~'];
  const atextArb = buildLowerAlphaNumericArbitrary(others);
  const localPartArb = adapter(
    // Maximal length for the output of dotMapper is 64,
    // In other words:
    // - `stringOf(atextArb, ...)` cannot produce values having more than 64 characters
    // - `array(...)` cannot produce more than 32 values
    array(
      stringOf(atextArb, {
        minLength: 1,
        maxLength: 64,
        size: constraints.size,
      }),
      { minLength: 1, maxLength: 32, size: constraints.size },
    ),
    dotAdapter,
  ).map(dotMapper, dotUnmapper);

  return tuple(localPartArb, domain({ size: constraints.size })).map(atMapper, atUnmapper);
}
