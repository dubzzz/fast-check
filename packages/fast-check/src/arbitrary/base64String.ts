import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import { base64 } from './base64';
import { MaxLengthUpperBound } from './_internals/helpers/MaxLengthFromMinLength';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
import { stringToBase64Mapper, stringToBase64Unmapper } from './_internals/mappers/StringToBase64';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For base64 strings
 *
 * A base64 string will always have a length multiple of 4 (padded with =)
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
function base64String(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  const { minLength: unscaledMinLength = 0, maxLength: unscaledMaxLength = MaxLengthUpperBound, size } = constraints;

  const minLength = unscaledMinLength + 3 - ((unscaledMinLength + 3) % 4);
  const maxLength = unscaledMaxLength - (unscaledMaxLength % 4);
  const requestedSize = constraints.maxLength === undefined && size === undefined ? '=' : size;

  if (minLength > maxLength) throw new Error('Minimal length should be inferior or equal to maximal length');
  if (minLength % 4 !== 0) throw new Error('Minimal length of base64 strings must be a multiple of 4');
  if (maxLength % 4 !== 0) throw new Error('Maximal length of base64 strings must be a multiple of 4');

  const charArbitrary = base64();
  const experimentalCustomSlices = createSlicesForString(charArbitrary, codePointsToStringUnmapper);
  const enrichedConstraints: ArrayConstraintsInternal<string> = {
    minLength,
    maxLength,
    size: requestedSize,
    experimentalCustomSlices,
  };
  return array(charArbitrary, enrichedConstraints)
    .map(codePointsToStringMapper, codePointsToStringUnmapper)
    .map(stringToBase64Mapper, stringToBase64Unmapper);
}
export { base64String };
