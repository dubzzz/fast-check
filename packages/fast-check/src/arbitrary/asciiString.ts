import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { array } from './array';
import { ascii } from './ascii';
import { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
export { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For strings of {@link ascii}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
export function asciiString(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  return array(ascii(), constraints).map(codePointsToStringMapper, codePointsToStringUnmapper);
}
