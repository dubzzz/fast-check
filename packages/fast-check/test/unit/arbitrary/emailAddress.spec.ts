import fc from 'fast-check';
import type { EmailAddressConstraints } from '../../../src/arbitrary/emailAddress';
import { emailAddress } from '../../../src/arbitrary/emailAddress';
import { Value } from '../../../src/check/arbitrary/definition/Value';

import {
  assertProduceSameValueGivenSameSeed,
  assertProduceCorrectValues,
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
} from './__test-helpers__/ArbitraryAssertions';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';
import { relativeSizeArb, sizeArb } from './__test-helpers__/SizeHelpers';

function beforeEachHook() {
  jest.resetModules();
  jest.restoreAllMocks();
  fc.configureGlobal({ beforeEach: beforeEachHook });
}
beforeEach(beforeEachHook);

describe('emailAddress (integration)', () => {
  const expectValidEmailRfc1123 = (t: string) => {
    // Taken from https://www.w3.org/TR/html5/forms.html#valid-e-mail-address
    const rfc1123 =
      /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
    expect(t).toMatch(rfc1123);
  };

  const expectValidEmailRfc2821 = (t: string) => {
    const [localPart, domain] = t.split('@');
    // The maximum total length of a user name or other local-part is 64 characters.
    expect(localPart.length).toBeLessThanOrEqual(64);
    // The maximum total length of a domain name or number is 255 characters.
    expect(domain.length).toBeLessThanOrEqual(255);
  };

  const expectValidEmailRfc5322 = (t: string) => {
    // Taken from https://stackoverflow.com/questions/201323/how-to-validate-an-email-address-using-a-regular-expression
    const rfc5322 =
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    expect(t).toMatch(rfc5322);
  };

  type Extra = EmailAddressConstraints;
  const extraParameters: fc.Arbitrary<Extra> = fc.record(
    { size: fc.oneof(sizeArb, relativeSizeArb) },
    { requiredKeys: [] },
  );

  const isCorrect = (t: string) => {
    expectValidEmailRfc1123(t);
    expectValidEmailRfc2821(t);
    expectValidEmailRfc5322(t);
  };

  const emailAddressBuilder = () => emailAddress();

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(emailAddressBuilder, { extraParameters });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(emailAddressBuilder, isCorrect, { extraParameters });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(emailAddressBuilder, { extraParameters });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(emailAddressBuilder, { extraParameters });
  });

  it.each`
    rawValue
    ${'me@domain.com'}
  `('should be able to shrink $rawValue', ({ rawValue }) => {
    // Arrange
    const arb = emailAddress();
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
