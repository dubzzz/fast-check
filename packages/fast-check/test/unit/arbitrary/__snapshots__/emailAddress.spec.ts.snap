// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emailAddress (integration) should be able to shrink me@domain.com 1`] = `
""me@domain.com"
├> "e@domain.com"
|  ├> "a@domain.com"
|  |  ├> "a@aomain.com"
|  |  |  ├> "a@a.com"
|  |  |  |  ├> "a@a.om"
|  |  |  |  |  ├> "a@a.am"
|  |  |  |  |  |  ├> "a@a.aa"
|  |  |  |  |  |  ├> "a@a.ag"
|  |  |  |  |  |  |  ├> "a@a.ad"
|  |  |  |  |  |  |  |  └> "a@a.ac"
|  |  |  |  |  |  |  |     └> "a@a.ab"
|  |  |  |  |  |  |  |        └> "a@a.aa"
|  |  |  |  |  |  |  └> "a@a.af"
|  |  |  |  |  |  |     └> "a@a.ae"
|  |  |  |  |  |  |        └> "a@a.ad"
|  |  |  |  |  |  |           ├> "a@a.aa"
|  |  |  |  |  |  |           └> "a@a.ac"
|  |  |  |  |  |  |              └> "a@a.ab"
|  |  |  |  |  |  |                 └> "a@a.aa"
|  |  |  |  |  |  ├> "a@a.aj"
|  |  |  |  |  |  |  └> "a@a.ai"
|  |  |  |  |  |  |     └> "a@a.ah"
|  |  |  |  |  |  |        └> "a@a.ag"
|  |  |  |  |  |  |           ├> "a@a.aa"
|  |  |  |  |  |  |           ├> "a@a.ad"
|  |  |  |  |  |  |           |  └> "a@a.ac"
|  |  |  |  |  |  |           |     └> "a@a.ab"
|  |  |  |  |  |  |           |        └> "a@a.aa"
|  |  |  |  |  |  |           └> "a@a.af"
|  |  |  |  |  |  |              └> "a@a.ae"
|  |  |  |  |  |  |                 └> "a@a.ad"
|  |  |  |  |  |  |                    ├> "a@a.aa"
|  |  |  |  |  |  |                    └> "a@a.ac"
|  |  |  |  |  |  |                       └> "a@a.ab"
|  |  |  |  |  |  |                          └> "a@a.aa"
|  |  |  |  |  |  └> "a@a.al"
|  |  |  |  |  |     └> "a@a.ak"
|  |  |  |  |  |        └> "a@a.aj"
|  |  |  |  |  |           ├> "a@a.aa"
|  |  |  |  |  |           ├> "a@a.af"
|  |  |  |  |  |           |  ├> "a@a.ad"
|  |  |  |  |  |           |  |  └> "a@a.ac"
|  |  |  |  |  |           |  |     └> "a@a.ab"
|  |  |  |  |  |           |  |        └> "a@a.aa"
|  |  |  |  |  |           |  └> "a@a.ae"
|  |  |  |  |  |           |     └> "a@a.ad"
|  |  |  |  |  |           |        ├> "a@a.aa"
|  |  |  |  |  |           |        └> "a@a.ac"
|  |  |  |  |  |           |           └> "a@a.ab"
|  |  |  |  |  |           |              └> "a@a.aa"
|  |  |  |  |  |           ├> "a@a.ah"
|  |  |  |  |  |           |  └> "a@a.ag"
|  |  |  |  |  |           |     └> "a@a.af"
|  |  |  |  |  |           |        ├> "a@a.aa"
|  |  |  |  |  |           |        ├> "a@a.ad"
|  |  |  |  |  |           |        |  └> "a@a.ac"
|  |  |  |  |  |           |        |     └> "a@a.ab"
|  |  |  |  |  |           |        |        └> "a@a.aa"
|  |  |  |  |  |           |        └> "a@a.ae"
|  |  |  |  |  |           |           └> "a@a.ad"
|  |  |  |  |  |           |              ├> "a@a.aa"
|  |  |  |  |  |           |              └> "a@a.ac"
|  |  |  |  |  |           |                 └> "a@a.ab"
|  |  |  |  |  |           |                    └> "a@a.aa"
|  |  |  |  |  |           └> "a@a.ai"
|  |  |  |  |  |              └> "a@a.ah"
|  |  |  |  |  |                 ├> "a@a.aa"
|  |  |  |  |  |                 ├> "a@a.ae"
|  |  |  |  |  |                 |  ├> "a@a.ac"
|  |  |  |  |  |                 |  |  └> "a@a.ab"
|  |  |  |  |  |                 |  |     └> "a@a.aa"
|  |  |  |  |  |                 |  └> "a@a.ad"
|  |  |  |  |  |                 |     └> "a@a.ac"
|  |  |  |  |  |                 |        ├> "a@a.aa"
|  |  |  |  |  |                 |        └> "a@a.ab"
|  |  |  |  |  |                 |           └> "a@a.aa"
|  |  |  |  |  |                 └> "a@a.ag"
|  |  |  |  |  |                    └> "a@a.af"
|  |  |  |  |  |                       └> "a@a.ae"
|  |  |  |  |  |                          ├> "a@a.aa"
|  |  |  |  |  |                          ├> "a@a.ac"
|  |  |  |  |  |                          |  └> "a@a.ab"
|  |  |  |  |  |                          |     └> "a@a.aa"
|  |  |  |  |  |                          └> "a@a.ad"
|  |  |  |  |  |                             └> "a@a.ac"
|  |  |  |  |  |                                ├> "a@a.aa"
|  |  |  |  |  |                                └> "a@a.ab"
|  |  |  |  |  |                                   └> "a@a.aa"
|  |  |  |  |  ├> "a@a.hm"
|  |  |  |  |  |  ├> "a@a.em"
|  |  |  |  |  |  |  ├> "a@a.cm"
|  |  |  |  |  |  |  |  ├> "a@a.bm"
|  |  |  |  |  |  |  |  |  ├> "a@a.am"
|  |  |  |  |  |  |  |  |  |  ├> "a@a.aa"
|  |  |  |  |  |  |  |  |  |  ├> "a@a.ag"
|  |  |  |  |  |  |  |  |  |  |  ├> "a@a.ad"
|  |  |  |  |  |  |  |  |  |  |  |  └> "a@a.ac"
|  |  |  |  |  |  |  |  |  |  |  |     └> "a@a.ab"
|  |  |  |  |  |  |  |  |  |  |  |        └> "a@a.aa"
|  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
